sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1305,11 +1305,20 @@ object TransposeWindow extends Rule[LogicalPlan] {
     }
   }
 
+  // Returns false if the parent 'w1' window function order by any of the window
+  // expressions produced by 'w2'.
+  private def compatibleOrderBy(w1: Window, w2: Window): Boolean = {
+    val childWindowExprs = w2.windowExpressions.map(x => x.toAttribute)
+    val parentOrder = w1.orderSpec.flatMap(x => x.references)
+    childWindowExprs.intersect(parentOrder).isEmpty
+  }
+
   private def windowsCompatible(w1: Window, w2: Window): Boolean = {
     w1.references.intersect(w2.windowOutputSet).isEmpty &&
```

Comment:
I'm a bit confused. Isn't the ORDER BY column part of `Window#references`? The code here should prevent this bug from happening IIUC.

Reply:
It's a subtle bug.
`references` by definition do not include output attributes (see https://github.com/apache/spark/blob/798fce3b571907ee52058004cc38c2e8dbc4b016/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala#L91).
So if the top Window function order by C and projects (outputs) C, then C is not included into references.
Therefore, we need that extra check.

(and we can't just take all attributes in the Window operator either, since PARTITION BY clauses do need to intersect)

Reply:
no, the `Window` operator won't exclude C, as it only excludes `windowOutputSet`